Fix #153/#154 : MultipleChoice/ItemSelection Interaction View #188

veena14cs opened the pull request:
This PR is duplicate of #[188](https://github.com/oppia/oppia-android/pull/188). It contains UI structure for Multiple choice and item selection interactions, and fixes #153 and #154.

Reference Design Doc: https://docs.google.com/document/d/1HFwDFcqenAXepbYr5rfawGHvkeBZpYvNMs228AWLn24/edit?usp=sharing

BenHenning committed:
Introduce first pass interface for ExplorationProgressController.

BenHenning committed:
Fill in the stubbed logic for ExplorationProgressController. Still no
tests to verify correctness.

Also, added a method to facilitate notifying of DataProvider changes on
the UI thread.

veena14cs committed:
added interactions

veena14cs committed:
Update ContentInteractionFragmentPresenter.kt

veena14cs committed:
Update ContentInteractionFragmentPresenter.kt

veena14cs committed:
updated urlImageParser file  name

BenHenning committed:
Fix lateinit issue in ExplorationProgressController due to wrongly ordered
initialization.

veena14cs committed:
created customviews. moved implementation to statefragment

veena14cs committed:
removed usused layout files

veena14cs committed:
fixed const values, fixed constructor

veena14cs committed:
removed extra spaces

veena14cs committed:
Update StateFragmentPresenter.kt

veena14cs committed:
updated todo

veena14cs committed:
removed initviews method

veena14cs committed:
reverted changes

BenHenning committed:
Fix a variaty of issues in the exp progress controller, properly hook it up to
the data controller, and start adding tests.

BenHenning committed:
Merge branch 'develop' into introduce-exploration-progress-controller

Conflicts:
	domain/build.gradle
	domain/src/main/java/org/oppia/domain/exploration/ExplorationDataController.kt

Also, migrate the data controller to the retriever.

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update state_fragment.xml

veena14cs committed:
Update StateFragmentPresenterTest.kt

veena14cs committed:
added test cases

rt4914 commented on utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:
```diff
@@ -0,0 +1,61 @@
+  /***
+   * Get the Drawable from URL
+   * @param urlString
+   * @return
```
Nit: describe the `@param` field.
This function is returning Drawable, so please mention that.

veena14cs replied:
described @param field and mentioned return field

rt4914 commented on utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:
```diff
@@ -0,0 +1,61 @@
+import com.bumptech.glide.request.transition.Transition
+import java.net.URL
+
+/** UrlImage Parser for android TextView to load Html Image tag.  */
```
There is extra space after `... tag.` remove that.

veena14cs replied:
fixed this java doc comment

rt4914 commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt:
```diff
@@ -0,0 +1,74 @@
+        override fun perform(uiController: androidx.test.espresso.UiController?, view: View?) {
+          //The view which we got in argument is the same view which Espresso found using onView(withId(R.id.parent_container))
+          val parentLinearLayout = view as LinearLayout
+//          //Get the checkbox inside the inner LinearLayout
```
Remove this comment.

veena14cs replied:
removed this comment

rt4914 commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt:
```diff
@@ -0,0 +1,74 @@
+          return "Performing click"
+        }
+        override fun perform(uiController: androidx.test.espresso.UiController?, view: View?) {
+          //The view which we got in argument is the same view which Espresso found using onView(withId(R.id.parent_container))
```
Remove this comment, if not needed.

veena14cs replied:
removed this comment

rt4914 commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt:
```diff
@@ -0,0 +1,74 @@
+        override fun perform(uiController: androidx.test.espresso.UiController?, view: View?) {
+          //The view which we got in argument is the same view which Espresso found using onView(withId(R.id.parent_container))
+          val parentRadioGroup = view as RadioGroup
+          //Get the LinearLayout inside the LinearLayout
```
Ditto as below comment.

veena14cs replied:
removed this comments

rt4914 commented on app/src/main/res/layout/state_fragment.xml:
```diff
@@ -29,6 +30,23 @@
+      android:layout_height="wrap_content"
+      android:layout_margin="10dp"
+      android:background="@color/white" android:orientation="vertical"
+      android:padding="10dp"
```
DP should be in multiple of 4 or 8.

veena14cs replied:
updated dp's

rt4914 commented on app/src/main/res/layout/state_fragment.xml:
```diff
@@ -29,6 +30,23 @@
+      android:layout_width="match_parent"
+      android:layout_height="wrap_content"
+      android:layout_margin="10dp"
+      android:background="@color/white" android:orientation="vertical"
```
Single line code.

veena14cs replied:
corrected

rt4914 commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -31,4 +43,62 @@ class StateFragmentPresenter @Inject constructor(
   }
+
+  private fun showInputInteractions() {
+    //Todo(Veena): Remove static initialization and use values from constructor for gaeCustomizationArgsMap and interactionInstanceId
```
Correct this TODO format.

`// Todo(Veena): Remove static initialization and use values from constructor for gaeCustomizationArgsMap and interactionInstanceId.`

veena14cs replied:
removed this Todo as the implementation function was separated.

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -14,12 +15,12 @@ class ExplorationActivityPresenter @Inject constructor(private val activity: App
+  private fun getExplorationFragment(): StateFragment? {
+    return activity.supportFragmentManager.findFragmentById(R.id.exploration_fragment_placeholder) as StateFragment?
   }
 }
```
I am assuming that changes in this files are just to make sure that whatever you have worked on in `StateFragment` can be checked as soon as the reviewer runs the code.
If i am correct, then you can revert all these changes and rather add these as instruction in your PR detail, by doing that we have two advantages:
1. Reviewer will have to review one less file.
2. This code can directly be merged in develop.

veena14cs replied:
it was added for debug and pushed code by mistake without reverting. Its been reverted now

rt4914 commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -31,4 +43,62 @@ class StateFragmentPresenter @Inject constructor(
+    gaeCustomizationArgsMap?.put("choices", sampleData)
+    interactionInstanceId = "MultipleChoiceInput"
+
+    // Todo(veena): Keep the below code for actual implementation
```
Nit change: `// Todo(Veena): Keep the below code for actual implementation.`

veena14cs replied:
fixed Todo

rt4914 commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -31,4 +43,62 @@ class StateFragmentPresenter @Inject constructor(
     getStateViewModel().setAudioFragmentVisible(isVisible)
   }
+
+  private fun showInputInteractions() {
```
Change this function to directly accept data, rather than creating data in this function.
This will help us in final implementation and we can directly pass data to the function and then it will manage it on its own.

veena14cs replied:
created dummy method to get data and separated this function for implementation

rt4914 commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt:
```diff
@@ -0,0 +1,74 @@
+          return "Performing click"
+        }
+        override fun perform(uiController: androidx.test.espresso.UiController?, view: View?) {
+          //The view which we got in argument is the same view which Espresso found using onView(withId(R.id.parent_container))
```
Remove this if not needed.

veena14cs replied:
removed unwanted comments

rt4914 reviewed (commented):
PTAL suggested changes.

veena14cs committed:
rmoved debug code

veena14cs committed:
fixed all issues

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
added indentation

veena14cs committed:
Update state_fragment.xml

veena14cs committed:
removed spaces

rt4914 commented on utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:
```diff
@@ -0,0 +1,61 @@
+  /***
+   * Get the Drawable from URL
+   * @param urlString contains the source of the image
+   * @return urlDrawable
```
Check this link and update accordingly:
https://github.com/oppia/oppia-android/blob/develop/data/src/test/java/org/oppia/data/backends/api/MockQuestionPlayerService.kt

veena14cs replied:
updated thanks.

rt4914 reviewed (commented):
Generally LGTM

veena14cs committed:
Update UrlImageParser.kt

BenHenning committed:
Created a separate ExplorationRetriever, hooked up
AnswerClassificationController, and attempted to make
ExplorationProgressController thread-safe.

The thread-safety led to significant interface changes in the progress
controller, and led to discovering some issues with the mediator live data
approach to interop coroutines and LiveData. This locking mechanism will
need to change since the optimal solution requires resolving #90.

BenHenning committed:
Merge branch 'develop' into introduce-exploration-progress-controller

BenHenning commented on app/src/main/res/layout/state_fragment.xml:
```diff
@@ -15,6 +16,7 @@
     android:layout_width="match_parent"
     android:layout_height="match_parent">
     <LinearLayout
+      android:id="@+id/linearLayout"
```
Please follow ``snake_case`` for layout identifier names for consistency with other identifiers in the codebase.

Ditto elsewhere.

BenHenning replied:
Also regarding naming: can we name this something with more context? 'linear layout' is a really generic name that doesn't tell us _what_ should be contained within the layout. The name ``interactionContainer`` below is much clearer since it's indicating that the layout is a container for interactions.

veena14cs replied:
followed snake_case naming conventions with clear context

BenHenning commented on utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:
```diff
@@ -0,0 +1,61 @@
+import java.net.URL
+
+/** UrlImage Parser for android TextView to load Html Image tag. */
+class UrlImageParser(internal var tvContents: TextView, internal var context: Context) : Html.ImageGetter {
```
Can you add dedicated tests for this parser? If it makes things easier, consider sending out a PR first with the parser & its tests to get that checked in right away, and then you can base off of that branch to complete the rest of this.

veena14cs replied:
Separate PR #201 is been made for UrlImageParser.

BenHenning commented on utility/src/main/java/org/oppia/util/data/UrlImageParser.kt:
```diff
@@ -0,0 +1,61 @@
+/** UrlImage Parser for android TextView to load Html Image tag. */
+class UrlImageParser(internal var tvContents: TextView, internal var context: Context) : Html.ImageGetter {
+
+  var targets: ArrayList<BitmapTarget>? = null
```
Why do we need this list? I'm concerned that it's potentially leaking these bitmap targets.

veena14cs replied:
It was not used anywhere so removed it.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt:
```diff
@@ -0,0 +1,69 @@
+    val density = resources.displayMetrics.density
+    val paddingDp = (paddingPixel * density).toInt()
+    gravity = Gravity.LEFT
+    setTextColor(ContextCompat.getColor(context, R.color.oppiaDarkBlue))
```
Can we set all of these properties (including the button tints) in a view-specific layout file, instead? I'm not keen on programmatically setting view properties because it makes them hard to find, especially when settings are split between code and XML. I think it would also largely simplify this & the next view since none of this extra initialization work needs to happen in code.

You may need to override the setText method, though, to make the HTML bit work correctly.

Ditto below for the other custom view.

veena14cs replied:
@BenHenning If we need to create recyclerview to show list of checkbox or radio button, there is no need for custom views right?

veena14cs replied:
Specified attributes in layout files for CustomCheckbox and CustomRadioButtons.
As we have used recyclerview for interactions and this holds checkbox and radiobuttons, Custom views are removed.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt:
```diff
@@ -0,0 +1,69 @@
+    })
+  }
+
+  fun convertHtmlToString(rawResponse: Any?, rdbtn: View): Spanned {
```
Could we move this to a general-purpose utility with dedicated tests? It seems shared with the custom radio button.

veena14cs replied:
moved html parser code  to utility

BenHenning commented on app/src/main/java/org/oppia/app/player/state/CustomCheckbox.kt:
```diff
@@ -0,0 +1,69 @@
+
+// TODO(#163): Move this to a StateCardFragment Low-fi PR.
+/** Custom Checkbox for MultipleSelectionInputInteractionView. */
+class CustomCheckbox(context: Context, private val optionContents: String) : CheckBox(context) {
```
Is it valid to have custom parameters in a view constructor? Since Android can construct these, I'd expect that if we want to pass in custom data we should instead use a setter method on the view.

BenHenning replied:
Instead of ``CustomCheckbox`` and ``CustomRadioButton`` maybe instead: ``HtmlCheckbox`` and ``HtmlRadioButton`` might be a bit clearer as to what these views are doing & how they should be used.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -1,26 +1,41 @@
   private val viewModelProvider: ViewModelProvider<StateViewModel>
 ) {
+
+  var gaeCustomizationArgsMap = HashMap<String, GaeCustomizationArgs>()
```
Here & elsewhere: please use the domain models defined in protos instead of GAE models for binding data to the UI. The domain models are set up with the intent of passing renderable data to the UI, whereas the GAE models are set up for the purpose of representing backend-provided data.

veena14cs replied:
used domain models to bind data

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -31,4 +46,62 @@ class StateFragmentPresenter @Inject constructor(
+    for (row in 0..0) {
+      for (i in 0..optionsArray.size - 1) {
+        val cb = CustomCheckbox(context, optionsArray[i])
+        binding.root.interactionContainer.addView(cb)
```
We should be adding views via a RecyclerView, instead. I think that using a RecyclerView + data-binding on the custom items will make it much easier to bind content to the custom checkbox/radio button, and will be closer to the final solution for this fragment.

veena14cs replied:
@BenHenning So I shall create separate adapters for these views?

veena14cs replied:
Added views vai RecyclerView. Removed custom views for interactions and added it in Interaction Adapter

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -31,4 +46,62 @@ class StateFragmentPresenter @Inject constructor(
+    gaeCustomizationArgsMap?.put("choices", sampleData)
+  }
+
+  private fun showInputInteractions(binding: StateFragmentBinding) {
```
Now that we have an ExplorationDataController, can we rely on that to provide data instead of faking it in this presenter? If you need some of the states in those test explorations to have more complex interactions/content, feel free to modify the [JSON](https://github.com/oppia/oppia-android/tree/develop/domain/src/main/assets) directly as needed.

veena14cs replied:
Use data from ExplorationDataController. It is updated in duplicate PR [#225](https://github.com/oppia/oppia-android/pull/225).

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt:
```diff
@@ -0,0 +1,81 @@
+import junit.framework.TestCase.assertEquals
+import org.hamcrest.Matcher
+
+/** Tests for [StateFragmentPresenter]. */
```
I think we should test presenters only through their fragments since they're more or less an implementation detail of the fragment. Could these tests instead go in StateFragmentTest?

veena14cs replied:
Implemented StateFragmentTest

nikitamarysolomanpvt commented on app/src/main/res/values/colors.xml:
```diff
@@ -13,4 +13,7 @@
   <!-- AUDIO COMPONENT -->
   <color name="audioComponentBackground">@color/oppiaDarkBlue</color>
+
+  <color name="blue_100">#0F0086FB</color>
```
We are following camel case is't?

veena14cs replied:
yes, but if there are more than one lighter shade of same color how to name it.

nikitamarysolomanpvt reviewed (approved):
LGTM

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt:
```diff
@@ -0,0 +1,81 @@
+    ActivityScenario.launch(ExplorationActivity::class.java).use {
+      interactionInstanceId = "MultipleChoiceInput"
+      assertEquals(interactionInstanceId, "MultipleChoiceInput")
+      onView(withId(R.id.interactionRadioGroup)).perform(object : ViewAction {
```
Could we use the built-in click() view action if we used a recycler view and something like: https://developer.android.com/training/testing/espresso/lists?

I think we should avoid custom ViewActions unless there isn't a good alternative.

veena14cs replied:
Used built-in click() of Recyclerview `onView(withId(R.id.rvInteractions)).perform(
        RecyclerViewActions.actionOnItemAtPosition<RecyclerView.ViewHolder>(2,click()))`

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/state/StateFragmentPresenterTest.kt:
```diff
@@ -0,0 +1,81 @@
+  var interactionInstanceId: String? = null
+
+  @Test
+  fun testMultipleSelectionInputInteraction() {
```
For here & below: please update these test names to include both the condition being tested and the expected outcome. See [this slide](https://docs.google.com/presentation/d/1jPyrVafvlCCaT0qY2r1Z1PYKL6MaklXo2QaO5FiJSKo/edit#slide=id.g62b813ee8d_0_134) for more context.

veena14cs replied:
Corrected test names as per the conventions

veena14cs replied:
@BenHenning based on `interactionInstanceId` I am checking wether it is MultipleChoice or ItemsSelection Input Interaction. While testing, I can only test one condition whichever is set in StateFragmentPresenter. If I want to test for both at a time, how to write a testcase for it.
For e.g I want testcase to behave as below
`interactionInstanceId = MultipleChoiceInput`
` - > runTestForMultiplechoice`
`interactionInstanceId = ItemsSelectionInput`
` - > runTestForItemSelection`

veena14cs replied:
This is updated in duplicate PR [#225](https://github.com/oppia/oppia-android/pull/225.)

BenHenning reviewed (changes requested):
Thanks @veena14cs! I had some high-level feedback. Please let me know if you'd like to discuss any of this in more detail.

veena14cs committed:
Removed custom views, used recyclerview for interactions

veena14cs committed:
worked on testcases

veena14cs committed:
Update StateFragmentTest.kt

veena14cs committed:
reverted debug code, moved htmlparser to utility

veena14cs committed:
Update codeStyleConfig.xml

veena14cs committed:
Merge branch 'introduce-exploration-progress-controller' into exploration-player-3-content-interaction

veena14cs committed:
working on fetching data from exploration

veena14cs committed:
working on fetching data from exploration

veena14cs committed:
revert manifest code

veena14cs commented:
Duplicate of this PR is been created [#225 ](https://github.com/oppia/oppia-android/pull/225).
